## Машинное обучение 1, ПМИ ФКН ВШЭ

## Практическое домашнее задание 1

### Общая информация

Дата выдачи: 25.09.2024

Мягкий дедлайн: 23:59MSK 06.10.2024

Жесткий дедлайн: 23:59MSK 10.10.2024

### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки `polars` и визуализации с помощью `matplotlib`. В первом разделе вам предстоит выполнить базовые задания с помощью вышеуказанных библиотек, а во втором распределить студентов по курсам. Баллы даются за выполнение отдельных пунктов. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотек polars и matplotlib.

### Оценивание и штрафы
Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов без учёта бонусов. Сдавать задание после жёсткого дедлайна нельзя.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

### Формат сдачи
Задания сдаются через систему Anytask. Инвайт можно получить у семинариста или ассистента. Присылать необходимо ноутбук с выполненным заданием. Сам ноутбук называйте в формате `homework-practice-01-polars-Username.ipynb`, где Username — ваша фамилия.

Для удобства проверки самостоятельно посчитайте свою максимальную оценку (исходя из набора решенных задач) и укажите ниже.

Оценка: xx.

## 0. Введение

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написанием production-кода.

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*.

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

In [1]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

4

Ячейка с неразмеченыным текстом.

Попробуйте создать свои ячейки, написать какой-нибудь код и текст какой-нибудь формулой.

In [2]:
# your code
print('\esokolov')

\esokolov


<>:2: SyntaxWarning: invalid escape sequence '\e'
<>:2: SyntaxWarning: invalid escape sequence '\e'
/var/folders/c1/2f76vwg555g_fgmfcx7m_j2w0000gn/T/ipykernel_19467/3899863378.py:2: SyntaxWarning: invalid escape sequence '\e'
  print('\esokolov')


[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. Выделять *текст* <s>при</s> **необходимости**
2. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

А ещё можно вставлять картинки, или гифки, или что захотите:

<img src="https://i.imgur.com/k5df8lX.gif" width="400px">



### Google Colab

Что за колаб?

**Google Colab (Colaboratory)** это **Jupyter Notebook + Cloud + Google Drive.**

Компания Google предоставляет возможность бесплатно запускать ноутбуки (предварительно загрузив их на свой гугл-диск) прямо в облаке. При этом вам не требуется установка никаких пакетов на свою машину, а работать можно прямиком из браузера. Вот ссылка:

https://colab.research.google.com

При использовании вы увидите много сходств с jupyter ноутбуком. Одним из преимуществ является доступность GPU, соответствующую опцию можно активировать в настройках сервиса. При желании вы сможете найти в интернете много туториалов по использованию или разобраться самостоятельно =)

### Kaggle

Альтернативой является **Kaggle**, он же площадка для самых разных соревнований по машинному обучению

Он, как и колаб, предоставляет бесплатный GPU, и даже не один, с чётким лимитом в 30 часов в неделю (colab может его внезапно отобрать, kaggle - нет) более мощные CPU с большим числом ядер, и больше Гб ОЗУ (раз так в 2.5)

https://www.kaggle.com

К сожалению мир не идеален, недостатки следующие:
- необходимость загружать датасеты через специальную кнопочку
- невозможность использовать свои скрипты и модули :(
- нужно тыкать техподдержку, чтобы создать аккаунт из России

## 1. Табличные данные и Polars

Эта версия домашки предполагает, что вы уже являетесь уверенным пользователем `pandas` и знакомы с датафреймами, как они устроены и как ими манипулировать. Библиотека `polars` основана на тех же концептах.

Главные достоинства, которые чуть позже продемонстрируем:
- он написан на Rust, поэтому почти все операции в нём выполняются значительно быстрее, есть распараллеливание из коробки, автоматическая оптимизация запросов и пр
- он позволяет работать с данными, которые не помещаются в оперативку, с этим поможет `pl.LazyFrame` и streaming
- конечно же имеются дополнительные функции, которых в `pandas` нет, но очень хочется, например агрегации с расширяющимся окном, произвольные оконные функции и прочее, для этого есть крайне гибкая штука `pl.Expr`

Подробнее можете прочитать на [любом](https://habr.com/ru/articles/710240/) [ресурсе](https://realpython.com/polars-python/), [посвящённом](https://blog.jetbrains.com/pycharm/2024/07/polars-vs-pandas/) [`polars`](https://habr.com/ru/companies/spectr/articles/738766/)

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами

Если не понятно, как вызывать конкретный метод, обратитесь к материалам ниже. Имейте в виду, что синтаксис `polars` отличается от привычного нампаевского.
Скорее всего все ваши самые извращённые фантазии уже как-то реализованы в `polars`, поэтому читайте внимательно. Ну и конечно гуглить не запрещено, но желательно прикладывать источник

Материалы:
1. [Getting started](https://docs.pola.rs/user-guide/getting-started/#reading-writing) из официального руководства
2. [Документация](https://docs.pola.rs/api/python/stable/reference/index.html) со всем необходимым
3. [Модификация для Data Science](https://polars-ds-extension.readthedocs.io/en/latest/index.html), вам скорее всего не понадобится в данном задании, но имейте в виду

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал polars и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно)

In [3]:
# поларс относительно молодой, имейте в виду, что что-то может сломаться, если будете менять версию
!pip3 install -qU hvplot xlsx2csv polars==1.5.0

In [4]:
import polars as pl

Скачаем данные:

In [5]:
!wget  -O 'end_seminar.xlsx' -q 'https://www.dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=0'

Для пользователей Windows: скачайте файл самостоятельно и поместите его в папку с тетрадкой. Или попробуйте один из следующих вариантов:

In [6]:
# !powershell iwr -outf somefile https://somesite/somefile

In [7]:
# !pip install wget
# import wget
# wget.download('https://dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=1', 'end_seminar.xlsx')

### 📌 **Обязательно прочитайте блок ниже**

> Пандасом и нампаем пользоваться неельзяя... Совсем

> В первой части задания (до раздела "Распределение студентов по курсам") использование циклов запрещается и повлечет за собой снижение оценки. Во второй уже можно, но с оговорками, о чём ниже

> Циклами в данном случае считается `for`, `while` и map-методы `polars`, где можно использовать лямбда-функцию. Главным образом `map_elements` и его производные.
И как правило, всё то, что не умеет делать `pl.LazyFrame` - список [здесь](https://docs.pola.rs/api/python/stable/reference/lazyframe/index.html). Мотивация следующая - используя встроенные функции и методы, `polars` может оптимизировать ваш запрос и провести все нужные манипуляции, прежде чем сериализовать и вывести ответ (за некоторыми исключениями), поэтому важно научиться с этим управляться

> Чтобы сделать выбор `polars` осмысленным, мы немножко модифицируем исходный датасет. Он будет больше в несколько раз и может не влезть в память, если у вас меньше 8 Гб, а с учётом преобразований может потребоваться до 64 Гб! Не пренебрегайте возможностями `pl.LazyFrame`. Его использование не обязательно, но может сильно облегчить вам жизнь

> Трансформировать датасет `df` можно, как только это потребуется по условию (таких пунктов всего 2), на всякий случай поставим значок 🔄

> Если небрежно обращаться с методами `polars`, вы рискуете крашнуть ядро, взорвать компутер, потерять рассудок... Если вдруг у вас есть тачка или хотите подебажить на исходном датасете, то пожалуйста, но помните про условия выше

> Гарантируется, что задание можно выполнить, имея лишь ресурсы Google Colab (бесплатные), удачи ( ͡° ͜ʖ ͡°)

Для каждой задачи из этого раздела вы должны написать код для получения ответа, а также дать текстовый ответ, если он предполагается.

На некоторые вопросы вы можете получить путём пристального взгляда на таблицу, но это не будет засчитываться. Вы в любом случае должны получить ответ с помощью кода.

### 🐻 Бонус polars. [3 балла]
Бонус за использование `polars` равен вашей оценке за дз, умноженной на коэффициент 1.3, поэтому даже если сделали не всё, не отчаивайтесь)

#### 0. Ещё пара красивых слов

Первые несколько манипуляций мы сделаем за вас, а заодно поясним, в чём тут нюанс

In [780]:
# читаем датафрейм
initial_df = pl.read_excel("end_seminar.xlsx", engine="xlsx2csv")
print(initial_df.shape[0], "rows")
print(initial_df.estimated_size("mb"), "mb")

361 rows
0.14590072631835938 mb


Как видим, он достаточно маленький. С такими объёмами выбор библиотеки это чисто вкусовщина. Давайте немного поднимем ставки

In [781]:
def repeat_df(df: pl.DataFrame | pl.LazyFrame, n: int = 2) -> pl.LazyFrame:
    return (
        df.lazy()
        .select(pl.col("timestamp").unique().sort().repeat_by(n).flatten())
        .join(df.lazy(), on="timestamp", how="left")
    )

Для работы с датафреймами, которые в память не влезают, нам понадобится объект `pl.LazyFrame`. Он представляет из себя схему выполнения вашего запроса, который оптимизируется, бьётся на батчи и параллелится под капотом через StreamingAPI, что для нас просто идеально. Перевести объект туда можно через метод `lazy`, обратно через `collect`. Учтите, что `streaming=True` работает не со всеми операциями, применяйте с умом, это можно проверить через `pl.LazyFrame.explain(streaming=True)`

In [782]:
repeat_n = 2

repeated_df = repeat_df(initial_df, repeat_n)
print("Schema")
print(repeated_df.explain(streaming=True, optimized=True))

Schema
LEFT JOIN:
LEFT PLAN ON: [col("timestamp")]
   SELECT [col("timestamp").unique().sort(asc).repeat_by([dyn int: 2]).explode()] FROM
    STREAMING:
      DF ["timestamp", "id", "rating", "18_group"]; PROJECT 1/16 COLUMNS; SELECTION: None
RIGHT PLAN ON: [col("timestamp")]
  DF ["timestamp", "id", "rating", "18_group"]; PROJECT */16 COLUMNS; SELECTION: None
END LEFT JOIN


In [783]:
# как видим, повторы работают
repeated_df = repeated_df.collect()
assert repeated_df.shape[0] == initial_df.shape[0] * repeat_n
print(repeated_df.shape[0], "rows")
print(repeated_df.estimated_size("mb"), "mb")

722 rows
0.2917900085449219 mb


Теперь ещё интереснее. Если бы наш датасет был не в 2 раза больше, а в 100000 раз, что бы мы делали? К счастью для нас, поларс может уместить 300кк строк всего лишь в 512Мб, и это даже не прикол, но работать с этим придётся немного иначе

In [784]:
N = 100000

df = repeat_df(initial_df, N)
print(initial_df.estimated_size("gb") * N, "gb")

14.248117804527283 gb


Поэтому даже такая тривиальная задача, как подсчёт строк становится уже не такой простой. Про это и будет дз

In [785]:
try:
    print(df.shape[0], "rows")
except AttributeError:
    print("ой, не получилось...")

# df.collect().shape[0] # тоже можно, если влезет
# len(df.collect().to_pandas()) # а такое может крашнуть ядро, попробуйте, хехе...

ой, не получилось...


In [786]:
def lazy_len(df: pl.LazyFrame) -> int:
    return df.select(pl.len()).collect().item()


rows = lazy_len(df)  # намного экономнее, потому что оптимизировано под капотом
assert rows == initial_df.shape[0] * N
print(rows, "rows")

36100000 rows


#### 1. [0.5 баллов] Откройте файл с таблицей (не забудьте про её формат). Выведите 3 последних строки

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы? (на это не надо отвечать, просто подумайте об этом)

In [787]:
df.columns

/var/folders/c1/2f76vwg555g_fgmfcx7m_j2w0000gn/T/ipykernel_19467/3289350417.py:1: PerformanceWarning: Determining the column names of a LazyFrame requires resolving its schema, which is a potentially expensive operation. Use `LazyFrame.collect_schema().names()` to get the column names without this warning.
  df.columns


['timestamp',
 'id',
 'rating',
 '18_group',
 'is_mi',
 'fall_1',
 'fall_2',
 'fall_3',
 'spring_1',
 'spring_2',
 'spring_3',
 'is_first_time',
 'percentile',
 '17_group',
 'blended',
 'is_ml_student']

In [788]:
# your code here
df.tail(3).collect()

shape: (3, 16)
┌─────────────┬─────────────┬────────┬──────────┬───┬────────────┬──────────┬─────────┬────────────┐
│ timestamp   ┆ id          ┆ rating ┆ 18_group ┆ … ┆ percentile ┆ 17_group ┆ blended ┆ is_ml_stud │
│ ---         ┆ ---         ┆ ---    ┆ ---      ┆   ┆ ---        ┆ ---      ┆ ---     ┆ ent        │
│ str         ┆ str         ┆ i64    ┆ i64      ┆   ┆ f64        ┆ str      ┆ str     ┆ ---        │
│             ┆             ┆        ┆          ┆   ┆            ┆          ┆         ┆ bool       │
╞═════════════╪═════════════╪════════╪══════════╪═══╪════════════╪══════════╪═════════╪════════════╡
│ 2020-05-24  ┆ 9d19b0d5f4f ┆ 695    ┆ 188      ┆ … ┆ 0.197309   ┆ null     ┆ null    ┆ null       │
│ 01:23:31    ┆ c8d7edc2258 ┆        ┆          ┆   ┆            ┆          ┆         ┆            │
│             ┆ 406f872c…   ┆        ┆          ┆   ┆            ┆          ┆         ┆            │
│ 2020-05-24  ┆ 9d19b0d5f4f ┆ 695    ┆ 188      ┆ … ┆ 0.197309   ┆ null     ┆ null    ┆ null       │
│ 01:23:31    ┆ c8d7edc2258 ┆        ┆          ┆   ┆            ┆          ┆         ┆            │
│             ┆ 406f872c…   ┆        ┆          ┆   ┆            ┆          ┆         ┆            │
│ 2020-05-24  ┆ 9d19b0d5f4f ┆ 695    ┆ 188      ┆ … ┆ 0.197309   ┆ null     ┆ null    ┆ null       │
│ 01:23:31    ┆ c8d7edc2258 ┆        ┆          ┆   ┆            ┆          ┆         ┆            │
│             ┆ 406f872c…   ┆        ┆          ┆   ┆            ┆          ┆         ┆            │
└─────────────┴─────────────┴────────┴──────────┴───┴────────────┴──────────┴─────────┴────────────┘

### Ответьте на вопросы:
1. Сколько было уникальных пользователей из групп 18-го года набора, а сколько из групп 17-го года?

In [789]:
df.filter(pl.col("18_group").is_not_null()).select(pl.col("id").unique()).count().collect()['id'][0]

213

In [790]:
df.filter(pl.col("17_group").is_not_null()).select(pl.col("id").unique()).count().collect()['id'][0]

134

2. Есть ли уникальные студенты с равными перцентилями?

In [791]:
# your code here
df.group_by("percentile").agg(pl.col("id").unique().alias("unique_ids")).filter(pl.col("unique_ids").list.len() > 1).collect()


shape: (1, 2)
┌────────────┬─────────────────────────────────┐
│ percentile ┆ unique_ids                      │
│ ---        ┆ ---                             │
│ f64        ┆ list[str]                       │
╞════════════╪═════════════════════════════════╡
│ 0.0        ┆ ["3e106aaef8b24dbb01319c73e14c… │
└────────────┴─────────────────────────────────┘

#### 2. [0.5 балла] Есть ли в данных пропуски? В каких колонках? Сколько их в каждой из этих колонок?

Уникальных или нет - не столь важно

In [792]:
# your code here
df.null_count().collect()

shape: (1, 16)
┌───────────┬─────┬────────┬──────────┬───┬────────────┬──────────┬──────────┬───────────────┐
│ timestamp ┆ id  ┆ rating ┆ 18_group ┆ … ┆ percentile ┆ 17_group ┆ blended  ┆ is_ml_student │
│ ---       ┆ --- ┆ ---    ┆ ---      ┆   ┆ ---        ┆ ---      ┆ ---      ┆ ---           │
│ u32       ┆ u32 ┆ u32    ┆ u32      ┆   ┆ u32        ┆ u32      ┆ u32      ┆ u32           │
╞═══════════╪═════╪════════╪══════════╪═══╪════════════╪══════════╪══════════╪═══════════════╡
│ 0         ┆ 0   ┆ 0      ┆ 13800000 ┆ … ┆ 0          ┆ 22300000 ┆ 22300000 ┆ 30400000      │
└───────────┴─────┴────────┴──────────┴───┴────────────┴──────────┴──────────┴───────────────┘

🔄 Преобразуйте датасет - заполните пропуски пустой строкой для строковых колонок, нулём для числовых и False для булевых (постарайтесь избежать перечисления названий всех столбцов).

In [793]:
# your code here
df = df.fill_null(strategy="zero")

In [794]:
df.null_count().collect()

shape: (1, 16)
┌───────────┬─────┬────────┬──────────┬───┬────────────┬──────────┬─────────┬───────────────┐
│ timestamp ┆ id  ┆ rating ┆ 18_group ┆ … ┆ percentile ┆ 17_group ┆ blended ┆ is_ml_student │
│ ---       ┆ --- ┆ ---    ┆ ---      ┆   ┆ ---        ┆ ---      ┆ ---     ┆ ---           │
│ u32       ┆ u32 ┆ u32    ┆ u32      ┆   ┆ u32        ┆ u32      ┆ u32     ┆ u32           │
╞═══════════╪═════╪════════╪══════════╪═══╪════════════╪══════════╪═════════╪═══════════════╡
│ 0         ┆ 0   ┆ 0      ┆ 0        ┆ … ┆ 0          ┆ 0        ┆ 0       ┆ 0             │
└───────────┴─────┴────────┴──────────┴───┴────────────┴──────────┴─────────┴───────────────┘

#### 3. [0.5 балла] Посмотрите повнимательнее на колонку 'is_first_time'.

Каково процентное соотношение ответов? Сколько из них "Нет"?

In [795]:
result = (
    df
    .select([
        pl.col('is_first_time').count().alias('total_responses'),
        pl.col('is_first_time').filter(pl.col('is_first_time') == 'Нет').count().alias('no_count'),
        pl.col('is_first_time').filter(pl.col('is_first_time') == 'Да').count().alias('yes_count')
    ])
    .collect()
)

total_responses = result['total_responses'][0]
no_count = result['no_count'][0]
yes_count = result['yes_count'][0]
no_percentage = (no_count / total_responses) * 100
yes_percentage = (yes_count/total_responses)*100

no_percentage, yes_percentage, 100 - yes_percentage - no_percentage

(14.127423822714682, 85.3185595567867, 0.5540166204986221)

🔄 Оставьте только самую позднюю версию обращений студентов и сохраните изменения

In [796]:
# your code here
df = df.sort('timestamp').unique(subset='id', keep='last')

In [797]:
lazy_len(df)

347

In [798]:
df.group_by("18_group").agg(pl.count().alias("count")).sort("count").collect()

/var/folders/c1/2f76vwg555g_fgmfcx7m_j2w0000gn/T/ipykernel_19467/493617960.py:1: DeprecationWarning: `pl.count()` is deprecated. Please use `pl.len()` instead.
  df.group_by("18_group").agg(pl.count().alias("count")).sort("count").collect()


shape: (9, 2)
┌──────────┬───────┐
│ 18_group ┆ count │
│ ---      ┆ ---   │
│ i64      ┆ u32   │
╞══════════╪═══════╡
│ 184      ┆ 21    │
│ 188      ┆ 23    │
│ 182      ┆ 26    │
│ 181      ┆ 27    │
│ 186      ┆ 27    │
│ 185      ┆ 28    │
│ 187      ┆ 30    │
│ 183      ┆ 31    │
│ 0        ┆ 134   │
└──────────┴───────┘

#### 4. [0.5 балла] Какие  blended-курсы для четверокурсников существуют? На какой blended-курс записалось наибольшее количество студентов? На каком из курсов собрались студенты с самым высоким средним рейтингом?

In [799]:
result = (
    df.filter(pl.col("blended") != "")
    .group_by("blended")
    .agg(pl.count().alias("count"))
    .sort("count", descending=True)
    .collect()
)
result

/var/folders/c1/2f76vwg555g_fgmfcx7m_j2w0000gn/T/ipykernel_19467/46966078.py:4: DeprecationWarning: `pl.count()` is deprecated. Please use `pl.len()` instead.
  .agg(pl.count().alias("count"))


shape: (3, 2)
┌─────────────────────────────────┬───────┐
│ blended                         ┆ count │
│ ---                             ┆ ---   │
│ str                             ┆ u32   │
╞═════════════════════════════════╪═══════╡
│ DevOps                          ┆ 87    │
│ Соревновательный анализ данных  ┆ 32    │
│ Введение в дифференциальную ге… ┆ 15    │
└─────────────────────────────────┴───────┘

In [800]:
df.filter(pl.col("blended") != "").group_by("blended").agg(pl.col('rating').mean().alias('mean_rating')).sort("mean_rating", descending=True).collect()

shape: (3, 2)
┌─────────────────────────────────┬─────────────┐
│ blended                         ┆ mean_rating │
│ ---                             ┆ ---         │
│ str                             ┆ f64         │
╞═════════════════════════════════╪═════════════╡
│ Введение в дифференциальную ге… ┆ 1210.733333 │
│ DevOps                          ┆ 1154.172414 │
│ Соревновательный анализ данных  ┆ 1144.21875  │
└─────────────────────────────────┴─────────────┘

#### 5. [1 балл] Выясните, есть ли студенты с абсолютно одинаковыми предпочтениями по всем курсам.

Для этого сформируйте таблицу, где для каждого возможного набора курсов посчитано количество студентов, выбравших такой набор, и оставьте только строки где это количество больше 1.

В данном случае набор курсов задается упорядоченным множеством ('fall_1', 'fall_2', 'fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'). Элемент blended будет нулевым для 3-го курса.

In [801]:
# your code here
df.group_by('fall_1', 'fall_2', 'fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended').agg(pl.col("id").count().alias("student_count")).filter(pl.col("student_count") > 1).collect()

shape: (15, 8)
┌────────────┬────────────┬────────────┬────────────┬────────────┬───────────┬─────────┬───────────┐
│ fall_1     ┆ fall_2     ┆ fall_3     ┆ spring_1   ┆ spring_2   ┆ spring_3  ┆ blended ┆ student_c │
│ ---        ┆ ---        ┆ ---        ┆ ---        ┆ ---        ┆ ---       ┆ ---     ┆ ount      │
│ str        ┆ str        ┆ str        ┆ str        ┆ str        ┆ str       ┆ str     ┆ ---       │
│            ┆            ┆            ┆            ┆            ┆           ┆         ┆ u32       │
╞════════════╪════════════╪════════════╪════════════╪════════════╪═══════════╪═════════╪═══════════╡
│ Безопаснос ┆ Теория баз ┆ Язык SQL   ┆ Машинное   ┆ Компьютерн ┆ Промышлен ┆         ┆ 2         │
│ ть компьют ┆ данных     ┆            ┆ обучение 2 ┆ ые сети    ┆ ное прогр ┆         ┆           │
│ ерных      ┆            ┆            ┆            ┆            ┆ аммирован ┆         ┆           │
│ сист…      ┆            ┆            ┆            ┆            ┆ ие …      ┆         ┆           │
│ Байесовски ┆ Анализ нес ┆ Сбор и     ┆ Обучение с ┆ Компьютерн ┆ Генератив ┆ DevOps  ┆ 2         │
│ е методы   ┆ труктуриро ┆ обработка  ┆ подкреплен ┆ ое зрение  ┆ ные       ┆         ┆           │
│ машинного  ┆ ванных     ┆ данных с   ┆ ием        ┆            ┆ модели в  ┆         ┆           │
│ о…         ┆ дан…       ┆ помо…      ┆            ┆            ┆ машинном… ┆         ┆           │
│ Statistica ┆ Анализ нес ┆ Сбор и     ┆ Обучение с ┆ Конфликты  ┆ Компьютер ┆ DevOps  ┆ 2         │
│ l Learning ┆ труктуриро ┆ обработка  ┆ подкреплен ┆ и          ┆ ное       ┆         ┆           │
│ Theory     ┆ ванных     ┆ данных с   ┆ ием        ┆ кооперация ┆ зрение    ┆         ┆           │
│            ┆ дан…       ┆ помо…      ┆            ┆            ┆           ┆         ┆           │
│ Безопаснос ┆ Сбор и     ┆ Матричные  ┆ Компьютерн ┆ Промышленн ┆ Анализ    ┆         ┆ 2         │
│ ть компьют ┆ обработка  ┆ вычисления ┆ ые сети    ┆ ое програм ┆ данных в  ┆         ┆           │
│ ерных      ┆ данных с   ┆            ┆            ┆ мирование  ┆ бизнесе   ┆         ┆           │
│ сист…      ┆ помо…      ┆            ┆            ┆ …          ┆           ┆         ┆           │
│ Байесовски ┆ Анализ нес ┆ Глубинное  ┆ Обучение с ┆ Генеративн ┆ Компьютер ┆ DevOps  ┆ 2         │
│ е методы   ┆ труктуриро ┆ обучение в ┆ подкреплен ┆ ые модели  ┆ ное       ┆         ┆           │
│ машинного  ┆ ванных     ┆ обработке… ┆ ием        ┆ в          ┆ зрение    ┆         ┆           │
│ о…         ┆ дан…       ┆            ┆            ┆ машинном…  ┆           ┆         ┆           │
│ …          ┆ …          ┆ …          ┆ …          ┆ …          ┆ …         ┆ …       ┆ …         │
│ Безопаснос ┆ Матричные  ┆ Моделирова ┆ Промышленн ┆ Компьютерн ┆ Дискретна ┆         ┆ 2         │
│ ть компьют ┆ вычисления ┆ ние        ┆ ое програм ┆ ые сети    ┆ я оптимиз ┆         ┆           │
│ ерных      ┆            ┆ временных  ┆ мирование  ┆            ┆ ация      ┆         ┆           │
│ сист…      ┆            ┆ рядов      ┆ …          ┆            ┆           ┆         ┆           │
│ Высокопрои ┆ Сбор и     ┆ Моделирова ┆ Проектиров ┆ Методы     ┆ Методы    ┆ DevOps  ┆ 2         │
│ зводительн ┆ обработка  ┆ ние        ┆ ание и     ┆ сжатия и   ┆ сжатия и  ┆         ┆           │
│ ые         ┆ данных с   ┆ временных  ┆ разработка ┆ передачи   ┆ передачи  ┆         ┆           │
│ вычисле…   ┆ помо…      ┆ рядов      ┆ вы…        ┆ медиа…     ┆ медиа…    ┆         ┆           │
│ Безопаснос ┆ Высокопрои ┆ Язык SQL   ┆ Компьютерн ┆ Промышленн ┆ Дискретна ┆         ┆ 3         │
│ ть компьют ┆ зводительн ┆            ┆ ые сети    ┆ ое програм ┆ я оптимиз ┆         ┆           │
│ ерных      ┆ ые         ┆            ┆            ┆ мирование  ┆ ация      ┆         ┆           │
│ сист…      ┆ вычисле…   ┆            ┆            ┆ …          ┆           ┆         ┆           │
│ Безопаснос ┆ Statistica ┆ Матричные  ┆ Дискр

In [802]:
df.select(pl.col("17_group")).unique().collect()

shape: (8, 1)
┌──────────┐
│ 17_group │
│ ---      │
│ str      │
╞══════════╡
│ 172      │
│ 171      │
│ 176      │
│ 177      │
│ 174      │
│          │
│ 175      │
│ 173      │
└──────────┘

#### 6. [0.5 балла] Найдите курсы по выбору, на которые записывались как студенты 18-го года набора, так и студенты 17-го года.

In [803]:
# your code here
courses18 = df.filter(pl.col("18_group") > 0).select([
          pl.col("fall_1"),
          pl.col("fall_2"),
          pl.col("fall_3"),
          pl.col("spring_1"),
          pl.col("spring_2"),
          pl.col("spring_3"),
          pl.col("blended")
      ]).melt().select('value').unique().collect()

courses17 = df.filter(pl.col("17_group") != "").select([
          pl.col("fall_1"),
          pl.col("fall_2"),
          pl.col("fall_3"),
          pl.col("spring_1"),
          pl.col("spring_2"),
          pl.col("spring_3"),
          pl.col("blended")
      ]).melt().select('value').unique().collect()

/var/folders/c1/2f76vwg555g_fgmfcx7m_j2w0000gn/T/ipykernel_19467/195554287.py:10: DeprecationWarning: `LazyFrame.melt` is deprecated. Use `unpivot` instead, with `index` instead of `id_vars` and `on` instead of `value_vars`
  ]).melt().select('value').unique().collect()
/var/folders/c1/2f76vwg555g_fgmfcx7m_j2w0000gn/T/ipykernel_19467/195554287.py:20: DeprecationWarning: `LazyFrame.melt` is deprecated. Use `unpivot` instead, with `index` instead of `id_vars` and `on` instead of `value_vars`
  ]).melt().select('value').unique().collect()


In [804]:
courses18.join(courses17, on="value", how="inner")

shape: (7, 1)
┌─────────────────────────────────┐
│ value                           │
│ ---                             │
│ str                             │
╞═════════════════════════════════╡
│ Численные методы                │
│ Принятие решений в условиях ри… │
│ Сбор и обработка данных с помо… │
│ Statistical Learning Theory     │
│ Безопасность компьютерных сист… │
│ Высокопроизводительные вычисле… │
│ Моделирование временных рядов   │
└─────────────────────────────────┘

Методом исключения найдите курсы, которые предлагались только студентам 18-го года и только студентам 17-го года.

In [805]:
# your code here
courses18.join(courses17, on="value", how="anti")

shape: (11, 1)
┌─────────────────────────────────┐
│ value                           │
│ ---                             │
│ str                             │
╞═════════════════════════════════╡
│ Системы баз данных              │
│ Компьютерные сети               │
│ Промышленное программирование … │
│ Машинное обучение 2             │
│ Анализ данных в бизнесе         │
│ …                               │
│ Язык SQL                        │
│                                 │
│ Теория баз данных               │
│ Дополнительные главы прикладно… │
│ Матричные вычисления            │
└─────────────────────────────────┘

In [806]:
courses17.join(courses18, on="value", how="anti")

shape: (13, 1)
┌─────────────────────────────────┐
│ value                           │
│ ---                             │
│ str                             │
╞═════════════════════════════════╡
│ Анализ неструктурированных дан… │
│ Введение в дифференциальную ге… │
│ DevOps                          │
│ Методы сжатия и передачи медиа… │
│ Обучение с подкреплением        │
│ …                               │
│ Конфликты и кооперация          │
│ Генеративные модели в машинном… │
│ Проектирование и разработка вы… │
│ Глубинное обучение в обработке… │
│ Байесовские методы машинного о… │
└─────────────────────────────────┘

### Визуализации

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. Здесь будут описаны ключевые правила оформления графиков для **всех** домашних заданий.

У matplotlib, конечно же, есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторой величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить один из них. Не забывайте про базовые принципы построения приличных графиков:
- оси должны быть подписаны, причём не слишком мелко;
- у графика должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если возможно сделать график интерактивным, это обычно того стоит - помогает решить проблемы с читаемостью;
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах);
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать;
- если распределение на гистограмме имеет тяжёлые хвосты, лучше использовать логарифмическую шкалу.

Еще одна билиотека для визуализации: [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Это настройка над matplotlib, иногда удобнее и красивее делать визуализации через неё.

У `polars` уже есть встроенные инструменты для визуализации, в частности интерактивные на базе [hvplot](https://docs.pola.rs/user-guide/misc/visualization/#plotly). Из её преимуществ - хорошая производительность при работе с большими данными, есть поддержка самых популярных видов графиков, но синтаксис не самый удобный

#### 7. [0.5 балла] Постройте график средних рейтингов по дням получения ответов (bar plot).

In [807]:
!pip3 install pyarrow

In [808]:
average_ratings = df.with_columns(
    pl.col('timestamp').str.strptime(pl.Date, "%Y-%m-%d %H:%M:%S")
)

average_ratings = (
    average_ratings.group_by("timestamp")
    .agg(pl.col("rating").mean().alias("average_rating"))
    .sort("timestamp")
)


In [809]:
average_ratings

In [810]:
import hvplot.polars

In [811]:
plot = average_ratings.hvplot.bar(
    x='timestamp',
    y='average_rating',
    width=650,
    height=400,
    title='Средние рейтинги по дням',
    xlabel='Дата',
    ylabel='Средний рейтинг',
    color='orange', 
    alpha=0.7,     
    legend='top_right', 
    grid=True,   
    tools=['hover']
)
plot

:Bars   [timestamp]   (average_rating)

Сохраните график в формате pdf (так он останется векторизованным) или html (тогда останется интерактивным)

In [812]:
# your code here
hvplot.save(plot, 'average_ratings_plot.html')

### 2. Распределение студентов по курсам.

<span style="color:red">!!!ВНИМАТЕЛЬНО ИЗУЧИТЕ ТЕКСТ НИЖЕ!!!</span>.

<span style="color:orange">Если во время выполнения заданий у вас вознинут вопросы -- еще раз перечитайте текст целиком, скорее всего ответы уже содержатся в нем.</span>

Теперь вам нужно распределить студентов по осенним курсам по выбору, учитывая их предпочтения.

## 📌 **Алгоритм распределения студентов по курсам**:
1. По умолчанию на каждой дисциплине по выбору у 3 и 4 курсов может учиться 1 группа (до 30 студентов). Исключения описаны ниже. На blended-дисциплинах для четверокурсников количество мест не ограничено.
2. Проводится первая волна отбора. Для каждой дисциплины формируется список тех, кто указал её первым приоритетом (если студент должен выбрать два курса по выбору, то для него дисциплины, которые он указал первым и вторым приоритетом, рассматриваются как дисциплины первого приоритета). Если желающих больше, чем мест, то выбирается топ по перцентилю рейтинга.
3. На дисциплинах, где остались места после первой волны, формируются списки тех, кто выбрал их вторым приоритетом, и места заполняются лучшими по перцентили рейтинга студентами. После этого проводится такая же процедура для дисциплин третьего приоритета.
4. Если студент не попал на необходимое количество курсов по итогам трёх волн, с ним связывается учебный офис и решает вопрос в индивидуальном порядке.

## ☝️ **Обращаем ваше внимание на следующие детали:**

- Конкурс на каждый курс общий для 3-го и 4-го курса

- По умолчанию студент выбирает один осенний и один весенний курс по выбору, а также четверокурсники выбирают один blended-курс. Студенты групп 17-го года специализаций МОП и ТИ выбирают по 2 осенних и 2 весенних курса по выбору, также студенты групп 18' специализации МИ выбирают 2 осенних курса. <i>Для студентов, которые выбирают 2 курса (например, осенних) первый приоритет — <code>fall_1</code> и <code>fall_2</code>, второй приоритет — <code>fall_3</code>. Такие студенты участвуют только в двух волнах отбора</i>.

- Студенты специализации МОП не могут выбрать весенним курсом по выбору Машинное обучение 2. **Если студент специализации МОП выбрал Машинное обучение 2, то его приоритеты сдвигаются.** Из-за совпадений первого и второго курса по выбору двигать приоритеты не надо.

- Blended-курсы не трогайте, по ним не надо распределять, на другие курсы они никак не влияют.

- Заведомо известно, что в процессе распределения не возникнет ситуации, когда на одно место претендуют студенты с одинаковым перцентилем.

- Постарайтесь воздержаться от использования циклов там, где это возможно. <i>Допустимо итерироваться по <b>курсам</b>, на которые проводится отбор, и по <b>волнам</b> отбора. Если вы придумаете, как обойтись и без этих циклов, то на усмотрение проверяющего могут быть добавлены бонусные баллы. <b>Дублирование кода не признается успешным избавлением от циклов</b></i>

- На выходе ожидается файл res_fall.csv с результатами распределения на осенние курсы по выбору. Файл должен быть следующего формата:

    * три столбца: ID, course1, course2
    
    * Если студент не попал на курс, но должен был, то вместо названия курса в ячейке должна быть строка "???"
    
    * Если студент должен выбрать только один курс, то в колонке course2 для него должна стоять строка "-"

    * Если студент должен выбрать два курса по выбору, то порядок в колонках course1 и course2 не важен.
    
    * Формат csv: для сохранения воспользуйтесь `df.write_csv('solution.csv')`

Для работы вам могут понадобиться следующие данные:

- Результаты опроса (вы уже использовали этот файл в первой части задания, но на всякий случай ссылка: https://www.dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=0)

- Соответствие номеров групп специализациям:

    * 171, 172 - МОП; 173 - ТИ; 174 — АДИС; 175, 176 — РС; 177 — АПР
    
    * У студентов 18-го года номера групп соответствуют номерам до распределения по специализациям. Это означает, что по номеру группы 18* нельзя однозначно определить специализацию студента. При этом в рамках распределения важно знать информацию только о двух из них: МОП и МИ. Эти знание можно получить из колонок 'is_ml_student ' и 'is_mi' соответственно.

- Ограничения по количеству мест на курсах по выбору:

    * Осенние: везде 30 мест, кроме Statistical Learning Theory (60 мест), Высокопроизводительных вычислений (60 мест), Анализа неструктурированных данных ($\infty$ мест)

    * Весенние: везде 30 мест, кроме Обучения с подкреплением (60 мест), Анализа данных в бизнесе (60 мест).


Кстати, убедитесь, что в данных больше нет пропусков и повторных записей и сериализуйте датафрейм

#### 0. Проверка

Для начала давайте убедимся, что вы успешно выполнили задания первой части и проверим ваши данные на наличие пропусков и повторов:

In [814]:
df = df.collect(streaming=True)

assert df.shape[0] == 347, "В таблице остались повторы или потеряны данные"
assert df.null_count().sum_horizontal().item() == 0, "В таблице остались пропуски"

Если вы не получили AssertionError, то можете продолжать.

#### 1. [1 балл] Создайте новый признак, обозначающий, сколько осенних курсов должен выбрать студент

В этом вам может помочь информация о специализации и группе студента.

In [815]:
# (￣^￣)ゞ
df = df.with_columns(
    (pl.when(pl.col("17_group") != "")
        .then(
            pl.when(pl.col("17_group").is_in(["171", "172"]))
            .then(2)
            .when(pl.col("17_group") == "173")
            .then(2)
            .otherwise(1)
        )
        .otherwise(0) +
     pl.when(pl.col("18_group") != 0)
        .then(
            pl.when(pl.col("is_mi") == 1)
            .then(2)
            .otherwise(1)
        )
        .otherwise(0))
    .alias("fall_courses_required") 
)

Проверка:

In [816]:
col_name = "fall_courses_required"

ids = [
    "2662600c2c37e11e62f6ee0b88452f22",
    "d555d2805e1d93d4f023e57dc4c8f403",
    "8fe79f84f36e3a5d2d6745621321302c",
    "e4caca755ee0bdd711e18fb8084958b5",
]
assert (
    df.filter(pl.col("id").is_in(ids)).sort("id")[col_name] == [2, 1, 2, 1]
).all()

#### 2. [2 балла] Распределите студентов в соответствии с первым приоритетом

In [890]:
def update_course_capacity(df, semester, course_capacity):
    unique_courses = (
        df.select(pl.col(f"{semester}_1").alias("course"))
        .vstack(df.select(pl.col(f"{semester}_2").alias("course")))
        .vstack(df.select(pl.col(f"{semester}_3").alias("course")))
        .unique()
    )
    for course in unique_courses["course"]:
        if course not in course_capacity:
            course_capacity[course] = 30
            
    return course_capacity

course_capacity_fall = {
    "Statistical Learning Theory": 60,
    "Высокопроизводительные вычисления": 60,
    'Анализ неструктурированных данных': 1e8
}

course_capacity_fall = update_course_capacity(df, "fall", course_capacity_fall)
course_capacity_fall

{'Statistical Learning Theory': 60,
 'Высокопроизводительные вычисления': 60,
 'Анализ неструктурированных данных': 100000000.0,
 'Теория баз данных': 30,
 'Сбор и обработка данных с помощью краудсорсинга': 30,
 'Матричные вычисления': 30,
 'Принятие решений в условиях риска и неопределённости': 30,
 'Безопасность компьютерных систем': 30,
 'Язык SQL': 30,
 'Моделирование временных рядов': 30,
 'Глубинное обучение в обработке звука': 30,
 'Байесовские методы машинного обучения': 30}

In [891]:
def mark_students(df: pl.DataFrame, course_capacity: dict) -> pl.DataFrame:
    result = {}

    for course in course_capacity.keys():
        df_1 = df.filter(((pl.col('fall_1') == course) | (pl.col('fall_2') == course)) & (pl.col('fall_courses_required') == 2))
        df_2 = df.filter((pl.col('fall_1') == course) & (pl.col('fall_courses_required') == 1))
        final_df = pl.concat([df_1, df_2])
        final_df = final_df.sort('percentile', descending=False)
        selected_students = final_df.head(min(course_capacity[course], len(final_df)))
        max_rating = selected_students['percentile'].min()
        min_rating = selected_students['percentile'].max()

        for student_id in selected_students['id']:
            student_rating = selected_students.filter(pl.col('id') == student_id)['percentile'].to_numpy()[0]
            is_first_place = True if student_rating == max_rating else None
            is_last_place = True if student_rating == min_rating else None
            
            if student_id not in result:
                result[student_id] = [is_first_place, is_last_place]
            else:
                result[student_id][0] = is_first_place if is_first_place is not None else result[student_id][0]
                result[student_id][1] = is_last_place if is_last_place is not None else result[student_id][1]


    df_not = df.filter(~(pl.col('id').is_in(result.keys())))
    for student_id in df_not['id']:
        result[student_id] = [None, None]

    results_df = pl.DataFrame({
        'id': list(result.keys()),
        'is_first_place': [value[0] for value in result.values()],
        'is_last_place': [value[1] for value in result.values()]
    })

    return results_df

df_wave_1 = mark_students(df, course_capacity_fall)

Здесь для проверки приведена таблица, в которой есть 2 дополнительные колонки:
    
    1) is_first_place - является ли студент лучшим по перцентили хотя бы на одном из курсов, куда он был зачислен
    (True / NaN)
    
    2) is_last_place  - является ли студент худшим по перцентили хотя бы на одном из курсов, куда он был зачислен (True / NaN)

In [819]:
!wget  -O '2_task_check.csv' -q 'https://www.dropbox.com/s/v8o2zzq3iz5gc9w/_2_task_check.csv?dl=0'
check_df = pl.read_csv('2_task_check.csv')

После распределения студентов в соответствии с первым приоритетом добавьте в свой датафрейм аналогичные признаки и запустите проверку:

In [892]:
fir_col_name =  "is_first_place"
last_col_name =  "is_last_place"

assert (
    check_df.sort("id").select("is_first_place", "is_last_place")
    == df_wave_1.sort("id").select(fir_col_name, last_col_name)
).sum().rows() == [(10, 12)]

#### 3. [2 балла] Проведите все три волны отбора студентов на курсы по выбору

Мне очень жалко удалять неоптимизированные версии, поэтому я ее тоже оставлю, чтобы были видны страдания.

Сделала общую функцию для распределения студентов select_students, остальные можно не смотреть (а не оптимальные и не оценивать пж)

In [1002]:
def select_students(df: pl.DataFrame, course_capacity: dict, time: str) -> pl.DataFrame:

    result_1 = pl.DataFrame(schema = {
    "id": str,
    "course": str}) 

    result_2 = pl.DataFrame(schema = {
    "id": str,
    "course": str}) 
    
    for course in course_capacity.keys():
        df_1 = df.filter(((pl.col(f'{time}_1') == course) | (pl.col(f'{time}_2') == course)) & (pl.col(f'{time}_courses_required') == 2))
        df_2 = df.filter((pl.col(f'{time}_1') == course) & (pl.col(f'{time}_courses_required') == 1))
        final_df = pl.concat([df_1, df_2]).sort('percentile', descending=False)

        selected_students = final_df.head(min(course_capacity[course], len(final_df)))
        course_capacity[course] -= len(selected_students)
        result_2 = pl.concat([result_2, selected_students.filter(pl.col(f'{time}_courses_required') == 2).select(['id', pl.lit(course).alias('course')])])
        result_1 = pl.concat([result_1, selected_students.filter(pl.col(f'{time}_courses_required') == 1).select(['id', pl.lit(course).alias('course')])])


    r_2 = result_2.group_by('id').agg(pl.count('course').alias('course_count')).filter(pl.col('course_count') < 2)['id'] #выделили челов с 2 курсами, которые недобрали курсы в первую волну
    unique_ids = result_1['id'].unique().to_list() + result_2['id'].unique().to_list() # это челы, которык набрали норм курсов себе

    df_not = pl.concat([df.filter(~pl.col('id').is_in(unique_ids)),
        df.filter(pl.col('id').is_in(r_2))]) # берем всех, не попавших никуда в первую волну и недобравших один курс


    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_1 = df_not.filter((pl.col(f'{time}_3') == course) & (pl.col(f'{time}_courses_required') == 2))
            df_2 = df_not.filter((pl.col(f'{time}_2') == course) & (pl.col(f'{time}_courses_required') == 1))
            df_1 = df_1.filter(~(pl.col(f'{time}_1') == course))
            final_df = pl.concat([df_1, df_2]).sort('percentile', descending=False)

            selected_students = final_df.head(min(len(final_df), course_capacity[course]))
            course_capacity[course] -= len(selected_students)

            result_2 = pl.concat([result_2, selected_students.filter(pl.col(f'{time}_courses_required') == 2).select(['id', pl.lit(course).alias('course')])])
            result_1 = pl.concat([result_1, selected_students.filter(pl.col(f'{time}_courses_required') == 1).select(['id', pl.lit(course).alias('course')])])

    df_not_2 = df.filter(~pl.col('id').is_in(result_1['id'].unique().to_list() + r_2.to_list()))

    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_2 = df_not_2.filter((pl.col(f'{time}_3') == course) & (pl.col(f'{time}_courses_required') == 1))
            final_df = df_2.sort('percentile', descending=False)

            selected_students = final_df.head(min(course_capacity[course], len(final_df)))
            course_capacity[course] -= len(selected_students)

            result_1 = pl.concat([result_1, selected_students.select(['id', pl.lit(course).alias('course')])])

    

    all_ids = pl.DataFrame(df.select('id'))
    
    result_2_list = result_2.group_by('id').agg(
    pl.col("course").unique().alias("course"))
    
    bad_1_course = df.filter((~(pl.col('id').is_in(result_1.select('id'))) & pl.col(f'{time}_courses_required') == 1)).select('id')

    res2 = result_2_list.select(
    pl.col('id'),
    pl.col('course').list.first().alias('course1'),
    pl.when(pl.col('course').list.len() == 2)
      .then(pl.col('course').list.last())
      .otherwise(pl.lit("???"))
      .alias('course2'))

    final_result = all_ids.join(
        result_1.with_columns(pl.lit('-').alias('course2')).rename({'course': 'course1'}),
        on='id',
        how='left'
    )
    final_result = final_result.join(res2, on='id', how = 'left')

    final_result = final_result.with_columns(
    pl.when(pl.col('course1_right').is_not_null()).then(pl.col('course1_right')).otherwise(pl.col('course1')).alias('course1'),
    pl.when(pl.col('course2_right').is_not_null()).then(pl.col('course2_right')).otherwise(pl.col('course2')).alias('course2'))

    final_result = final_result.drop(['course1_right', 'course2_right'])

    
    final_result = final_result.with_columns([
    pl.col('course1').fill_null('???'),
    pl.col('course2').fill_null('???')])

    final_result = final_result.with_columns(
        pl.when(pl.col('id').is_in(bad_1_course.select('id').unique()))
        .then(pl.lit('-'))
        .otherwise(pl.col('course2'))
        .alias('course2'))
    
    final_result = final_result.rename({"id": "ID"})
    return final_result

In [1012]:
def select_students_fall_not_optimal(df: pl.DataFrame, course_capacity: dict) -> pl.DataFrame:
    result_1 = {} # здесь челы с одним курсом по выбору
    result_2 = {} # здесь челы с двумя курсами по выбору

    # первая волна
    for course in course_capacity.keys():
        df_1 = df.filter(((pl.col('fall_1') == course) | (pl.col('fall_2') == course)) & (pl.col('fall_courses_required') == 2))
        df_2 = df.filter((pl.col('fall_1') == course) & (pl.col('fall_courses_required') == 1))
        final_df = pl.concat([df_1, df_2])
        final_df = final_df.sort('percentile', descending=False)
        selected_students = final_df.head(min(course_capacity[course], len(final_df)))
        course_capacity[course] -= len(selected_students)

        for student_id in selected_students['id']:
            required_courses = df.filter(pl.col('id') == student_id).select('fall_courses_required').to_numpy()[0]
            if required_courses == 2:
                if student_id not in result_2.keys():
                    result_2[student_id] = [course]
                else:
                    result_2[student_id].append(course)
            else:
                result_1[student_id] = [course]
            

    r_2 = []
    for i in result_2.keys():
        if len(set(result_2[i])) <2:
            r_2.append(i)
    

    # вторая волна
    df_not = pl.concat([df.filter(~pl.col('id').is_in(set(result_1.keys()).union(set(result_2.keys())))), df.filter(pl.col('id').is_in(set(r_2)))])
    
    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_1 = df_not.filter(((pl.col('fall_3') == course)) & (pl.col('fall_courses_required') == 2))
            df_2 = df_not.filter((pl.col('fall_2') == course) & (pl.col('fall_courses_required') == 1))
            df_1 = df_1.filter(~(pl.col('fall_1') == course))
            final_df = pl.concat([df_1, df_2])
            final_df = final_df.sort('percentile', descending=False)
            selected_students = final_df.head(min(len(final_df), course_capacity[course]))
            course_capacity[course] -= len(selected_students)

            for student_id in selected_students['id']:
                required_courses = df.filter(pl.col('id') == student_id).select('fall_courses_required').to_numpy()[0]
                if required_courses == 2:
                    if student_id not in result_2.keys():
                        result_2[student_id] = [course]
                    else:
                        result_2[student_id].append(course)
                else:
                    result_1[student_id] = [course]

    # третья волна только для челиков, выбирающих 1 курс
    df_not_2 = df.filter(~pl.col('id').is_in(set(result_1.keys()).union(set(r_2))))
    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_2 = df_not_2.filter((pl.col('fall_3') == course) & (pl.col('fall_courses_required') == 1))
            final_df = df_2.sort('percentile', descending=False)
            selected_students = final_df.head(min(course_capacity[course], len(final_df)))
            course_capacity[course] -= len(selected_students)

            for student_id in selected_students['id']:
                result_1[student_id] = [course]

    output_data = []
    all_results = {**result_1, **result_2}
    for student_id, courses in all_results.items():=
        required_courses = df.filter(pl.col('id') == student_id).select('fall_courses_required').to_numpy()[0]
        if len(set(courses)) == 0:
            if required_courses == 2:
                output_data.append((student_id, '???', '???'))
            elif required_courses == 1:
                output_data.append((student_id, '???', '-'))
        elif len(set(courses)) == 1:
            if required_courses == 2:
                output_data.append((student_id, courses[0], '???'))
            elif required_courses == 1:
                output_data.append((student_id, courses[0], '-'))
        else:
            output_data.append((student_id, courses[0], courses[1]))

    to_contact = df.filter(~pl.col('id').is_in(set(result_1.keys()).union(set(result_2.keys()))))
    for student_id in to_contact['id'].to_numpy():
        required_courses = df.filter(pl.col('id') == student_id).select('fall_courses_required').to_numpy()[0][0]
        
        if required_courses == 2:
            output_data.append((student_id, '???', '???'))
        elif required_courses == 1:
            output_data.append((student_id, '???', '-'))

    result_df = pl.DataFrame(output_data, schema=["ID", "course1", "course2"])

    return result_df


In [1004]:
def select_students_fall(df: pl.DataFrame, course_capacity: dict) -> pl.DataFrame:

    result_1 = pl.DataFrame(schema = {
    "id": str,
    "course": str}) 

    result_2 = pl.DataFrame(schema = {
    "id": str,
    "course": str}) 
    
    for course in course_capacity.keys():
        df_1 = df.filter(((pl.col('fall_1') == course) | (pl.col('fall_2') == course)) & (pl.col('fall_courses_required') == 2))
        df_2 = df.filter((pl.col('fall_1') == course) & (pl.col('fall_courses_required') == 1))
        final_df = pl.concat([df_1, df_2]).sort('percentile', descending=False)

        selected_students = final_df.head(min(course_capacity[course], len(final_df)))
        course_capacity[course] -= len(selected_students)
        result_2 = pl.concat([result_2, selected_students.filter(pl.col('fall_courses_required') == 2).select(['id', pl.lit(course).alias('course')])])
        result_1 = pl.concat([result_1, selected_students.filter(pl.col('fall_courses_required') == 1).select(['id', pl.lit(course).alias('course')])])


    r_2 = result_2.group_by('id').agg(pl.count('course').alias('course_count')).filter(pl.col('course_count') < 2)['id'] #выделили челов с 2 курсами, которые недобрали курсы в первую волну
    unique_ids = result_1['id'].unique().to_list() + result_2['id'].unique().to_list() # это челы, которык набрали норм курсов себе

    df_not = pl.concat([df.filter(~pl.col('id').is_in(unique_ids)),
        df.filter(pl.col('id').is_in(r_2))]) # берем всех, не попавших никуда в первую волну и недобравших один курс


    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_1 = df_not.filter((pl.col('fall_3') == course) & (pl.col('fall_courses_required') == 2))
            df_2 = df_not.filter((pl.col('fall_2') == course) & (pl.col('fall_courses_required') == 1))
            df_1 = df_1.filter(~(pl.col('fall_1') == course))
            final_df = pl.concat([df_1, df_2]).sort('percentile', descending=False)

            selected_students = final_df.head(min(len(final_df), course_capacity[course]))
            course_capacity[course] -= len(selected_students)

            result_2 = pl.concat([result_2, selected_students.filter(pl.col('fall_courses_required') == 2).select(['id', pl.lit(course).alias('course')])])
            result_1 = pl.concat([result_1, selected_students.filter(pl.col('fall_courses_required') == 1).select(['id', pl.lit(course).alias('course')])])

    df_not_2 = df.filter(~pl.col('id').is_in(result_1['id'].unique().to_list() + r_2.to_list()))

    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_2 = df_not_2.filter((pl.col('fall_3') == course) & (pl.col('fall_courses_required') == 1))
            final_df = df_2.sort('percentile', descending=False)

            selected_students = final_df.head(min(course_capacity[course], len(final_df)))
            course_capacity[course] -= len(selected_students)

            result_1 = pl.concat([result_1, selected_students.select(['id', pl.lit(course).alias('course')])])

    

    all_ids = pl.DataFrame(df.select('id'))
    
    result_2_list = result_2.group_by('id').agg(
    pl.col("course").unique().alias("course"))
    
    bad_1_course = df.filter((~(pl.col('id').is_in(result_1.select('id'))) & pl.col('fall_courses_required') == 1)).select('id')

    res2 = result_2_list.select(
    pl.col('id'),
    pl.col('course').list.first().alias('course1'),
    pl.when(pl.col('course').list.len() == 2)
      .then(pl.col('course').list.last())
      .otherwise(pl.lit("???"))
      .alias('course2'))

    final_result = all_ids.join(
        result_1.with_columns(pl.lit('-').alias('course2')).rename({'course': 'course1'}),
        on='id',
        how='left'
    )
    final_result = final_result.join(res2, on='id', how = 'left')

    final_result = final_result.with_columns(
    pl.when(pl.col('course1_right').is_not_null()).then(pl.col('course1_right')).otherwise(pl.col('course1')).alias('course1'),
    pl.when(pl.col('course2_right').is_not_null()).then(pl.col('course2_right')).otherwise(pl.col('course2')).alias('course2'))

    final_result = final_result.drop(['course1_right', 'course2_right'])

    
    final_result = final_result.with_columns([
    pl.col('course1').fill_null('???'),
    pl.col('course2').fill_null('???')])

    final_result = final_result.with_columns(
        pl.when(pl.col('id').is_in(bad_1_course.select('id').unique()))
        .then(pl.lit('-'))
        .otherwise(pl.col('course2'))
        .alias('course2'))
    
    final_result = final_result.rename({"id": "ID"})
    return final_result

In [1005]:
def update_course_capacity(df, semester, course_capacity):
    unique_courses = (
        df.select(pl.col(f"{semester}_1").alias("course"))
        .vstack(df.select(pl.col(f"{semester}_2").alias("course")))
        .vstack(df.select(pl.col(f"{semester}_3").alias("course")))
        .unique()
    )
    for course in unique_courses["course"]:
        if course not in course_capacity:
            course_capacity[course] = 30
            
    return course_capacity

course_capacity_fall = {
    "Statistical Learning Theory": 60,
    "Высокопроизводительные вычисления": 60,
    'Анализ неструктурированных данных': 1e8
}

course_capacity_fall = update_course_capacity(df, "fall", course_capacity_fall)
res = select_students(df, course_capacity_fall, 'fall')
res

shape: (347, 3)
┌──────────────────────────────────┬──────────────────────────────────┬────────────────────────────┐
│ ID                               ┆ course1                          ┆ course2                    │
│ ---                              ┆ ---                              ┆ ---                        │
│ str                              ┆ str                              ┆ str                        │
╞══════════════════════════════════╪══════════════════════════════════╪════════════════════════════╡
│ bba51e2f648cb3323b96858a404dac…  ┆ Моделирование временных рядов    ┆ Анализ неструктурированных │
│                                  ┆                                  ┆ дан…                       │
│ bfe75c08c6b7e31d4e7d48c9c5fed6…  ┆ Безопасность компьютерных сист…  ┆ -                          │
│ cabc5ea85cac3029dc2aecc4955f9c…  ┆ ???                              ┆ -                          │
│ 9808df02cd92ed45df39e6bc748f31…  ┆ Statistical Learning Theory      ┆ -                          │
│ 32fea7beb180e7de64ad408013cc01…  ┆ ???                              ┆ -                          │
│ …                                ┆ …                                ┆ …                          │
│ 5daee1bd98559bb2c52373e4169422…  ┆ Statistical Learning Theory      ┆ -                          │
│ f0f6f3881894172a0f04e79b554b49…  ┆ Принятие решений в условиях ри…  ┆ -                          │
│ abceddfa2fc4eae8cff4a4e712abb3…  ┆ Моделирование временных рядов    ┆ -                          │
│ fc2c60b8004aa91bc272df84628621…  ┆ Язык SQL                         ┆ -                          │
│ 6ed89e424179b0510e3f7c0cb7369a…  ┆ Моделирование временных рядов    ┆ -                          │
└──────────────────────────────────┴──────────────────────────────────┴────────────────────────────┘

**Отправьте свой файл res_fall.csv в контест (будет в ближайшее время, следите за новостями) и прикрепите/укажите ниже ваш никнейм и ссылку на успешную посылку.**

*Дисклеймер:*

Контест выдаётся для самопроверки. Если ваша посылка получила ОК, то код, скорее всего, правильный. Но при этом оценка всё равно может быть снижена в случае обнаружения неэффективностей или ошибок в коде. Если вы сдадите в AnyTask очевидно неработающий код или ноутбук без кода, но при этом в контест будет сдан корректный файл, то это будет расцениваться как плагиат.

На всякий случай просим вас сдать вместе с ноутбуком файл res_fall.csv в anytask

In [1006]:
res.write_csv("res_fall.csv")

### 4. [1 балл] Распределите таким же образом студентов еще и на весенние курсы по выбору.

Если ваш код был хорошо структурирован, то это не составит проблем.

Если вы выполнили это задание, сдайте среди прочего файл res_spring.csv в таком же формате, как и res_fall.csv.

Перед тем, как засунуть в ту же функцию код, сдвинем предпочтения для моп-ошибулек.

In [893]:
strange = df.filter(pl.col('is_ml_student') & ((pl.col('spring_1') == "Машинное обучение 2") | (pl.col('spring_2') == "Машинное обучение 2") | (pl.col('spring_3') == "Машинное обучение 2") ))
df = df.filter(~pl.col('id').is_in(strange.select('id')))
strange_1 = strange.filter(pl.col('spring_1') == "Машинное обучение 2")
strange_1 = strange_1.with_columns([
    pl.when(pl.col('spring_1') == "Машинное обучение 2").then(pl.col('spring_2')).otherwise(pl.col('spring_1')).alias('spring_1'),
    pl.when(pl.col('spring_2').is_not_null()).then(pl.col('spring_3')).otherwise(pl.col('spring_2')).alias('spring_2'),
    pl.lit("???").alias('spring_3')
])

strange_2 = strange.filter(pl.col('spring_2') == "Машинное обучение 2")
strange_2 = strange_2.with_columns([
    pl.when(pl.col('spring_2') == "Машинное обучение 2").then(pl.col('spring_3')).otherwise(pl.col('spring_2')).alias('spring_2'),
    pl.lit("???").alias('spring_3')
])

strange_3 = strange.filter(pl.col('spring_3') == "Машинное обучение 2")
strange_3 = strange_3.with_columns([
    pl.lit("???").alias('spring_3')
])

strange = pl.concat([strange_1, strange_2, strange_3])
df = pl.concat([df, strange])
df

shape: (347, 18)
┌────────────┬────────────┬────────┬──────────┬───┬────────────┬───────────┬───────────┬───────────┐
│ timestamp  ┆ id         ┆ rating ┆ 18_group ┆ … ┆ blended    ┆ is_ml_stu ┆ fall_cour ┆ spring_co │
│ ---        ┆ ---        ┆ ---    ┆ ---      ┆   ┆ ---        ┆ dent      ┆ ses_requi ┆ urses_req │
│ str        ┆ str        ┆ i64    ┆ i64      ┆   ┆ str        ┆ ---       ┆ red       ┆ uired     │
│            ┆            ┆        ┆          ┆   ┆            ┆ bool      ┆ ---       ┆ ---       │
│            ┆            ┆        ┆          ┆   ┆            ┆           ┆ i32       ┆ i32       │
╞════════════╪════════════╪════════╪══════════╪═══╪════════════╪═══════════╪═══════════╪═══════════╡
│ 2020-05-17 ┆ bba51e2f64 ┆ 1052   ┆ 0        ┆ … ┆ Соревноват ┆ false     ┆ 2         ┆ 2         │
│ 18:29:44   ┆ 8cb3323b96 ┆        ┆          ┆   ┆ ельный     ┆           ┆           ┆           │
│            ┆ 858a404dac ┆        ┆          ┆   ┆ анализ     ┆           ┆           ┆           │
│            ┆ …          ┆        ┆          ┆   ┆ данных     ┆           ┆           ┆           │
│ 2020-05-18 ┆ bfe75c08c6 ┆ 758    ┆ 182      ┆ … ┆            ┆ true      ┆ 1         ┆ 1         │
│ 17:22:56   ┆ b7e31d4e7d ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ 48c9c5fed6 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ …          ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│ 2020-05-18 ┆ cabc5ea85c ┆ 686    ┆ 182      ┆ … ┆            ┆ true      ┆ 1         ┆ 1         │
│ 21:53:48   ┆ ac3029dc2a ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ ecc4955f9c ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ …          ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│ 2020-05-17 ┆ 9808df02cd ┆ 645    ┆ 187      ┆ … ┆            ┆ false     ┆ 1         ┆ 1         │
│ 10:27:38   ┆ 92ed45df39 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ e6bc748f31 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ …          ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│ 2020-05-18 ┆ 32fea7beb1 ┆ 539    ┆ 185      ┆ … ┆            ┆ true      ┆ 1         ┆ 1         │
│ 18:55:40   ┆ 80e7de64ad ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ 408013cc01 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ …          ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│ …          ┆ …          ┆ …      ┆ …        ┆ … ┆ …          ┆ …         ┆ …         ┆ …         │
│ 2020-05-16 ┆ 5daee1bd98 ┆ 593    ┆ 182      ┆ … ┆            ┆ true      ┆ 1         ┆ 1         │
│ 13:02:24   ┆ 559bb2c523 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ 73e4169422 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ …          ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│ 2020-05-15 ┆ f0f6f38818 ┆ 673    ┆ 181      ┆ … ┆            ┆ true      ┆ 1         ┆ 1         │
│ 21:55:16   ┆ 94172a0f04 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ e79b554b49 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ …          ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│ 2020-05-18 ┆ abceddfa2f ┆ 664    ┆ 183      ┆ … ┆            ┆ true      ┆ 1         ┆ 1         │
│ 23:01:03   ┆ c4eae8cff4 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ a4e712abb3 ┆        ┆          ┆   ┆            ┆           ┆           ┆           │
│            ┆ …          ┆        ┆        

In [1013]:
def select_students_spring_not_optimal(df: pl.DataFrame, course_capacity: dict) -> pl.DataFrame:
    result_1 = {} # здесь челы с одним курсом по выбору
    result_2 = {} # здесь челы с двумя курсами по выбору

    # первая волна
    for course in course_capacity.keys():
        df_1 = df.filter(((pl.col('spring_1') == course) | (pl.col('spring_2') == course)) & (pl.col('spring_courses_required') == 2))
        df_2 = df.filter((pl.col('spring_1') == course) & (pl.col('spring_courses_required') == 1))
        final_df = pl.concat([df_1, df_2])
        final_df = final_df.sort('percentile', descending=False)
        selected_students = final_df.head(min(course_capacity[course], len(final_df)))
        course_capacity[course] -= len(selected_students)

        for student_id in selected_students['id']:
            required_courses = df.filter(pl.col('id') == student_id).select('spring_courses_required').to_numpy()[0]
            if required_courses == 2:
                if student_id not in result_2.keys():
                    result_2[student_id] = [course]
                else:
                    result_2[student_id].append(course)
            else:
                result_1[student_id] = [course]
            

    r_2 = []
    for i in result_2.keys():
        if len(set(result_2[i])) <2:
            r_2.append(i)
    

    # вторая волна
    df_not = pl.concat([df.filter(~pl.col('id').is_in(set(result_1.keys()).union(set(result_2.keys())))), df.filter(pl.col('id').is_in(set(r_2)))])
    
    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_1 = df_not.filter(((pl.col('spring_3') == course)) & (pl.col('spring_courses_required') == 2))
            df_2 = df_not.filter((pl.col('spring_2') == course) & (pl.col('spring_courses_required') == 1))
            df_1 = df_1.filter(~(pl.col('spring_1') == course))
            final_df = pl.concat([df_1, df_2])
            final_df = final_df.sort('percentile', descending=False)
            selected_students = final_df.head(min(len(final_df), course_capacity[course]))
            course_capacity[course] -= len(selected_students)

            for student_id in selected_students['id']:
                required_courses = df.filter(pl.col('id') == student_id).select('spring_courses_required').to_numpy()[0]
                if required_courses == 2:
                    if student_id not in result_2.keys():
                        result_2[student_id] = [course]
                    else:
                        result_2[student_id].append(course)
                else:
                    result_1[student_id] = [course]

    # третья волна только для челиков, выбирающих 1 курс
    df_not_2 = df.filter(~pl.col('id').is_in(set(result_1.keys()).union(set(r_2))))
    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_2 = df_not_2.filter((pl.col('spring_3') == course) & (pl.col('spring_courses_required') == 1))
            final_df = df_2.sort('percentile', descending=False)
            selected_students = final_df.head(min(course_capacity[course], len(final_df)))
            course_capacity[course] -= len(selected_students)

            for student_id in selected_students['id']:
                result_1[student_id] = [course]

    output_data = []
    all_results = {**result_1, **result_2}
    for student_id, courses in all_results.items():
        required_courses = df.filter(pl.col('id') == student_id).select('spring_courses_required').to_numpy()[0]
        if len(set(courses)) == 0:
            if required_courses == 2:
                output_data.append((student_id, '???', '???'))
            elif required_courses == 1:
                output_data.append((student_id, '???', '-'))
        elif len(set(courses)) == 1:
            if required_courses == 2:
                output_data.append((student_id, courses[0], '???'))
            elif required_courses == 1:
                output_data.append((student_id, courses[0], '-'))
        else:
            output_data.append((student_id, courses[0], courses[1]))

    to_contact = df.filter(~pl.col('id').is_in(set(result_1.keys()).union(set(result_2.keys()))))
    for student_id in to_contact['id'].to_numpy():
        required_courses = df.filter(pl.col('id') == student_id).select('spring_courses_required').to_numpy()[0][0]
        
        if required_courses == 2:
            output_data.append((student_id, '???', '???'))
        elif required_courses == 1:
            output_data.append((student_id, '???', '-'))

    result_df = pl.DataFrame(output_data, schema=["ID", "course1", "course2"])

    return result_df

In [1011]:
def select_students_spring(df: pl.DataFrame, course_capacity: dict) -> pl.DataFrame:

    result_1 = pl.DataFrame(schema = {
    "id": str,
    "course": str}) 

    result_2 = pl.DataFrame(schema = {
    "id": str,
    "course": str}) 
    
    for course in course_capacity.keys():
        df_1 = df.filter(((pl.col('spring_1') == course) | (pl.col('spring_2') == course)) & (pl.col('spring_courses_required') == 2))
        df_2 = df.filter((pl.col('spring_1') == course) & (pl.col('spring_courses_required') == 1))
        final_df = pl.concat([df_1, df_2]).sort('percentile', descending=False)

        selected_students = final_df.head(min(course_capacity[course], len(final_df)))
        course_capacity[course] -= len(selected_students)
        result_2 = pl.concat([result_2, selected_students.filter(pl.col('spring_courses_required') == 2).select(['id', pl.lit(course).alias('course')])])
        result_1 = pl.concat([result_1, selected_students.filter(pl.col('spring_courses_required') == 1).select(['id', pl.lit(course).alias('course')])])


    r_2 = result_2.group_by('id').agg(pl.count('course').alias('course_count')).filter(pl.col('course_count') < 2)['id'] #выделили челов с 2 курсами, которые недобрали курсы в первую волну
    unique_ids = result_1['id'].unique().to_list() + result_2['id'].unique().to_list() # это челы, которык набрали норм курсов себе

    df_not = pl.concat([df.filter(~pl.col('id').is_in(unique_ids)),
        df.filter(pl.col('id').is_in(r_2))]) # берем всех, не попавших никуда в первую волну и недобравших один курс


    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_1 = df_not.filter((pl.col('spring_3') == course) & (pl.col('spring_courses_required') == 2))
            df_2 = df_not.filter((pl.col('spring_2') == course) & (pl.col('spring_courses_required') == 1))
            df_1 = df_1.filter(~(pl.col('spring_1') == course))
            final_df = pl.concat([df_1, df_2]).sort('percentile', descending=False)

            selected_students = final_df.head(min(len(final_df), course_capacity[course]))
            course_capacity[course] -= len(selected_students)

            result_2 = pl.concat([result_2, selected_students.filter(pl.col('spring_courses_required') == 2).select(['id', pl.lit(course).alias('course')])])
            result_1 = pl.concat([result_1, selected_students.filter(pl.col('spring_courses_required') == 1).select(['id', pl.lit(course).alias('course')])])

    df_not_2 = df.filter(~pl.col('id').is_in(result_1['id'].unique().to_list() + r_2.to_list()))

    for course in course_capacity.keys():
        if course_capacity[course] > 0:
            df_2 = df_not_2.filter((pl.col('spring_3') == course) & (pl.col('spring_courses_required') == 1))
            final_df = df_2.sort('percentile', descending=False)

            selected_students = final_df.head(min(course_capacity[course], len(final_df)))
            course_capacity[course] -= len(selected_students)

            result_1 = pl.concat([result_1, selected_students.select(['id', pl.lit(course).alias('course')])])

    

    all_ids = pl.DataFrame(df.select('id'))
    
    result_2_list = result_2.group_by('id').agg(
    pl.col("course").unique().alias("course"))
    
    bad_1_course = df.filter((~(pl.col('id').is_in(result_1.select('id'))) & pl.col('spring_courses_required') == 1)).select('id')

    res2 = result_2_list.select(
    pl.col('id'),
    pl.col('course').list.first().alias('course1'),
    pl.when(pl.col('course').list.len() == 2)
      .then(pl.col('course').list.last())
      .otherwise(pl.lit("???"))
      .alias('course2'))

    final_result = all_ids.join(
        result_1.with_columns(pl.lit('-').alias('course2')).rename({'course': 'course1'}),
        on='id',
        how='left'
    )
    final_result = final_result.join(res2, on='id', how = 'left')

    final_result = final_result.with_columns(
    pl.when(pl.col('course1_right').is_not_null()).then(pl.col('course1_right')).otherwise(pl.col('course1')).alias('course1'),
    pl.when(pl.col('course2_right').is_not_null()).then(pl.col('course2_right')).otherwise(pl.col('course2')).alias('course2'))

    final_result = final_result.drop(['course1_right', 'course2_right'])

    
    final_result = final_result.with_columns([
    pl.col('course1').fill_null('???'),
    pl.col('course2').fill_null('???')])

    final_result = final_result.with_columns(
        pl.when(pl.col('id').is_in(bad_1_course.select('id').unique()))
        .then(pl.lit('-'))
        .otherwise(pl.col('course2'))
        .alias('course2'))
    
    final_result = final_result.rename({"id": "ID"})
    return final_result

In [1007]:
df = df.with_columns(
    (pl.when(pl.col("17_group") != "")
        .then(
            pl.when(pl.col("17_group").is_in(["171", "172", "173"]))
            .then(2)
            .otherwise(1) 
        )
        .otherwise(0) +
     pl.when(pl.col("18_group") != 0)
        .then(1)
        .otherwise(0)
    )
    .alias("spring_courses_required")
)

course_capacity_spring = {
    "Обучение с подкреплением": 60, 
    "Анализ данных в бизнесе" : 60
}

course_capacity_spring = update_course_capacity(df, "spring", course_capacity_spring)
res = select_students(df, course_capacity_spring, 'spring')
res

shape: (347, 3)
┌───────────────────────────────────┬───────────────────────────────────┬──────────────────────────┐
│ ID                                ┆ course1                           ┆ course2                  │
│ ---                               ┆ ---                               ┆ ---                      │
│ str                               ┆ str                               ┆ str                      │
╞═══════════════════════════════════╪═══════════════════════════════════╪══════════════════════════╡
│ bba51e2f648cb3323b96858a404dac…   ┆ Обучение с подкреплением          ┆ Методы сжатия и передачи │
│                                   ┆                                   ┆ медиа…                   │
│ bfe75c08c6b7e31d4e7d48c9c5fed6…   ┆ Дискретная оптимизация            ┆ -                        │
│ cabc5ea85cac3029dc2aecc4955f9c…   ┆ Дискретная оптимизация            ┆ -                        │
│ 9808df02cd92ed45df39e6bc748f31…   ┆ Машинное обучение 2               ┆ -                        │
│ 32fea7beb180e7de64ad408013cc01…   ┆ Системы баз данных                ┆ -                        │
│ …                                 ┆ …                                 ┆ …                        │
│ 5daee1bd98559bb2c52373e4169422…   ┆ Анализ данных в бизнесе           ┆ -                        │
│ f0f6f3881894172a0f04e79b554b49…   ┆ Численные методы                  ┆ -                        │
│ abceddfa2fc4eae8cff4a4e712abb3…   ┆ Промышленное программирование …   ┆ -                        │
│ fc2c60b8004aa91bc272df84628621…   ┆ Системы баз данных                ┆ -                        │
│ 6ed89e424179b0510e3f7c0cb7369a…   ┆ Анализ данных в бизнесе           ┆ -                        │
└───────────────────────────────────┴───────────────────────────────────┴──────────────────────────┘

In [1008]:
course_capacity_spring

{'Обучение с подкреплением': 0,
 'Анализ данных в бизнесе': 28,
 'Символьные вычисления': 20,
 'Конфликты и кооперация': 10,
 'Компьютерное зрение': 0,
 'Методы сжатия и передачи медиаданных': 2,
 'Дополнительные главы прикладной статистики': 15,
 'Генеративные модели в машинном обучении': 12,
 'Компьютерные сети': 0,
 'Системы баз данных': 13,
 '???': 30,
 'Машинное обучение 2': 0,
 'Промышленное программирование на языке Java': 0,
 'Численные методы': 14,
 'Проектирование и разработка высоконагруженных сервисов': 0,
 'Дискретная оптимизация': 0}

In [1009]:
res.write_csv("res_spring.csv")

**Отправьте свой файл res_spring.csv в контест (будет в ближайшее время, следите за новостями) и прикрепите/укажите ниже ваш никнейм и ссылку на успешную посылку.**

На всякий случай просим вас сдать вместе с ноутбуком файл res_spring.csv в anytask

Вставьте картинку, описывающую ваш опыт выполнения этого задания:

In [1016]:
<a href="https://ibb.co/xmwxHsS"><img src="https://i.ibb.co/z8CqJ5h/photo-2024-10-06-15-55-44.jpg" alt="photo-2024-10-06-15-55-44" border="0" /></a>

SyntaxError: invalid syntax (3114456059.py, line 1)

In [1017]:
<a href="https://ibb.co/8rLWXGV"><img src="https://i.ibb.co/X7rhV6T/photo-2024-10-06-15-55-47.jpg" alt="photo-2024-10-06-15-55-47" border="0" /></a>

SyntaxError: invalid syntax (2038395264.py, line 1)